src/canvasdrawer.js

```diff
@@ -0,0 +1,1151 @@
+            rect.y * $.pixelDensityRatio,
+            rect.width * $.pixelDensityRatio,
+            rect.height * $.pixelDensityRatio
+        );
+
+        context.restore();
+    }
+
+    /**
+     * @private
+     * @inner
+     * Fires the tile-drawing event.
+     *
+     */
+    _raiseTileDrawingEvent(tiledImage, context, tile, rendered){
+        /**
+         * This event is fired just before the tile is drawn giving the application a chance to alter the image.
+         *
+         * NOTE: This event is only fired when the 'canvas' drawer is being used
+         *
+         * @event tile-drawing
+         * @memberof OpenSeadragon.Viewer
+         * @type {object}
+         * @property {OpenSeadragon.Viewer} eventSource - A reference to the Viewer which raised the event.
+         * @property {OpenSeadragon.Tile} tile - The Tile being drawn.
+         * @property {OpenSeadragon.TiledImage} tiledImage - Which TiledImage is being drawn.
+         * @property {CanvasRenderingContext2D} context - The HTML canvas context being drawn into.
+         * @property {CanvasRenderingContext2D} rendered - The HTML canvas context containing the tile imagery.
+         * @property {?Object} userData - Arbitrary subscriber-defined object.
+         */
+        this.viewer.raiseEvent('tile-drawing', {
+            tiledImage: tiledImage,
+            context: context,
+            tile: tile,
+            rendered: rendered
+        });
+    }
+
+    /**
+     * @private
+     * @inner
+     * Clears the Drawer so it's ready to draw another frame.
+     *
+     */
+    _prepareNewFrame() {
+        var viewportSize = this._calculateCanvasSize();
+        if( this.canvas.width !== viewportSize.x ||
+            this.canvas.height !== viewportSize.y ) {
+            this.canvas.width = viewportSize.x;
+            this.canvas.height = viewportSize.y;
+            this._updateImageSmoothingEnabled(this.context);
+            if ( this.sketchCanvas !== null ) {
+                var sketchCanvasSize = this._calculateSketchCanvasSize();
+                this.sketchCanvas.width = sketchCanvasSize.x;
+                this.sketchCanvas.height = sketchCanvasSize.y;
+                this._updateImageSmoothingEnabled(this.sketchContext);
+            }
+        }
+        this._clear();
+    }
+
+    /**
+     * @private
+     * @inner
+     * @param {Boolean} useSketch Whether to clear sketch canvas or main canvas
+     * @param {OpenSeadragon.Rect} [bounds] The rectangle to clear
+     */
+    _clear(useSketch, bounds){
+        var context = this._getContext(useSketch);
+        if (bounds) {
+            context.clearRect(bounds.x, bounds.y, bounds.width, bounds.height);
+        } else {
+            var canvas = context.canvas;
+            context.clearRect(0, 0, canvas.width, canvas.height);
+        }
+    }
+
+    /**
+     * @private
+     * @inner
+     * Draws a TiledImage.
+     *
+     */
+    _drawTiles( tiledImage ) {
```

Comment:
I've just skimmed this function, assuming it's mostly a copy and paste from `TiledImage`. Let me know if there's anything specific I should be looking at.

Reply:
Correct, this is all old code, just in a new location.